perceval/backends/core/rocketchat.py

```diff
@@ -0,0 +1,338 @@
+
+        return RocketChatClient(self.url, self.user_id, self.api_token, self.max_items, self.archive,
+                                from_archive, self.ssl_verify, roomId=self.roomId)
+
+    def __get_or_fetch_user(self, user_id, username, name):
+        if user_id in self._users:
+            return self._users[user_id]
+
+        logger.debug("User %s not found on client cache; fetching it", user_id)
+
+        user = {
+            "username": username,
+            "name": name,
+        }
+
+        self._users[user_id] = user
+        return user
+
+
+class RocketChatClient(HttpClient):
+    """RocketChat API client.
+    Client for fetching information from the RocketChat server
+    using its REST API.
+    :param api_token: key needed to use the API
+    :param max_items: maximum number of items per request
+    :param archive: an archive to store/read fetched data
+    :param from_archive: it tells whether to write/read the archive
+    :param ssl_verify: enable/disable SSL verification
+    """
+    ROOM_HISTORY = 'channels.history'
+    ROOM_INFO = 'channels.info'
+    ROOM_ID = 'roomId'
+    AUTH_TOKEN = 'X-Auth-Token'
+    USER_ID = 'X-User-Id'
+    PCOUNT = "count"
+    POLDEST = "oldest"
+
+    def __init__(self, url, user_id, api_token, max_items=MAX_ITEMS, archive=None, from_archive=False,
+                 ssl_verify=True, roomId=None):
+        super().__init__(url, archive=archive, from_archive=from_archive, ssl_verify=ssl_verify)
+        self.api_token = api_token
+        self.max_items = max_items
+        self.roomId = roomId
+        self.user_id = user_id
+
+    def room_info(self, roomId):
+        """Fetch information about a room."""
+
+        resource = self.ROOM_INFO
+
+        params = {
+            self.ROOM_ID: roomId,
+        }
+
+        response = self._fetch(resource, params)
+
+        return response
+
+    def history(self, roomId, oldest=None):
+        """Fetch the history of a room."""
+
+        resource = self.ROOM_HISTORY
+
+        params = {
+            self.ROOM_ID: roomId,
+            self.PCOUNT: self.max_items,
+            self.POLDEST: "oldest",
+        }
+
+        response = self._fetch(resource, params)
+
+        return response
+
+    def _fetch(self, resource, params):
+        """Fetch a resource.
+        :param resource: resource to get
+        :param params: dict with the HTTP parameters needed to get
+            the given resource
+        """
+        url = urijoin(self.base_url, API_EXTENSION, resource)
+
+        headers = {
+            self.AUTH_TOKEN: self.api_token,
+            self.USER_ID: self.user_id
+        }
+
+        logger.debug("RocketChat client requests: %s params: %s",
+                     resource, str(params))
+
+        r = self.fetch(url, payload=params, headers=headers)
+
+        # Check for possible API errors
+        result = r.json()
+
+        if r.status_code != 200:
+            raise RocketChatClientError(error=result['error'])
```

Comment:
In which cases is the exception triggered?

Reply:
One example is `400 Client Error: Bad Request for url`, I'm not sure if others exist. Also, there are cases when the api returns an html page instead of a response, one case being if the server url is incorrect. There are also cases when the response status code is 200, but the api encountered an error and sent it inside the response body, like:

    {'success': False,
     'error': 'Invalid query parameter provided: "{"_updatedAt" : { "$gt" : { "date : 1526847048093}}}" [error-invalid-query]',
     'errorType': 'error-invalid-query',
     'details': {'helperMethod': 'parseJsonQuery'}}

I've added error handling for these too.